MetaBoxes: Remove dirty-checking metaboxes state

youknowriad opened the pull request:
closes #4183 and closes #3975 

Due to how MetaBoxes work it's not possible to dirty-check them properly. It's not possible to know if they contain default values as well. Thus, this PR removes the meta boxes dirty checking making Gutenberg work like the classic editor when there are active meta boxes which means:

 - The "save draft" and "update" button are never disabled
 - If you try to leave the editor's page with changes in the meta boxes area, you're not prompted with the confirmation box.

youknowriad committed:
MetaBoxes: Remove dirty-checking metaboxes state

spencerfinnell commented:
This allows some of my custom meta boxes to continue working as before. 👍

wsydney76 commented:
> you're not prompted with the confirmation box.

If that means it might happen to loose a ton of edited content by accident with one wrong click, i don't think that's an option.

If it is not possible to detect a dirty state correctly, i would instead prefer to be always prompted, even if there is no change at all.

youknowriad commented:
> If that means it might happen to loose a ton of edited content by accident with one wrong click, i don't think that's an option.

That's how the current editor works, and it doesn't seem to bother anyone. The confirmation box is still enabled for changes to the post content and the post settings though.

wsydney76 commented:
> That's how the current editor works

If so, that is bad and should be fixed.

>  it doesn't seem to bother anyone

Except all the users who lost their edits.

Using ACF Pro, there is obviously an inconsistent behaviour, changes in most field types cause a prompt,
but some do not (simple textfield for example). 

It should at least be documented what plugin authors have to do to cause a prompt (something like a setDirty() call?)

youknowriad commented:
> It should at least be documented what plugin authors have to do to cause a prompt (something like a setDirty() call?)

Sounds like a good idea to add an API like that. I think it's out of the scope of the current PR thought since this brings us to a similar behavior to the current Editor.

youknowriad commented:
Can I have some thoughts here @aduth @mtias

aduth commented:
>Due to how MetaBoxes work it's not possible to dirty-check them properly.

Is it possible to check whether there are "unsaved" changes during the `beforeunload` event? Unsaved meaning we serialize metabox forms when first loading, then compare serialized form values when user tries to leave?

(I should perhaps note I don't have the clearest understanding of the meta boxes behaviors)

That said, if it's prohibitively difficult, my preference would be toward resolving the bugs of #4183 and #3975 even if it means losing the dirtiness prompts.

youknowriad commented:
> Is it possible to check whether there are "unsaved" changes during the beforeunload event? Unsaved meaning we serialize metabox forms when first loading, then compare serialized form values when user tries to leave?

Good proposal, I might try it, it may be difficult to achieve though since the "state" of the metaboxes is dispatched through several areas and if the areas are not visible, the DOM is in a hidden node.

But it has a downside. Some metaboxes come with default values, even if they do not have any change, they are still "dirty" (needs to be saved)

aduth commented:
>Some metaboxes come with default values, even if they do not have any change, they are still "dirty" (needs to be saved)

I'd agree these are "dirty" only in that they need to be included in the save payload, but wouldn't agree that they'd need to stop the user from navigating away. If we're not already, couldn't we just send the entire metabox form of fields in the save request, whether or not any changes have been made?

youknowriad commented:
>  If we're not already, couldn't we just send the entire metabox form of fields in the save request, whether or not any changes have been made?

Yes, that's what I'm doing in this PR alongside the the removal of the dirty checking to prevent leaving the editor (which matchers the old editor's behavior). 

I'll try to update the PR to make it work similarily to what you propose. But the only way to make it behave properly is to avoid relying on the "onChange" event and check the state when the user wants to leave the page because we can't catch updates done programmatically with onChange events.

youknowriad committed:
MetaBoxes: compare meta boxes HTML while leaving the editor to warn about unsaved changes

youknowriad commented:
@aduth I added the check in 39364A0 I'm not totally satisfied with the code but can't think of a good alternative right now

aduth commented on editor/store/selectors.js:
```diff
@@ -351,7 +354,7 @@ export function isCurrentPostPublished( state ) {
 export function isEditedPostPublishable( state ) {
 	const post = getCurrentPost( state );
-	return isEditedPostDirty( state ) || [ 'publish', 'private', 'future' ].indexOf( post.status ) === -1;
+	return isEditedPostDirty( state ) || hasMetaBoxes( state ) || [ 'publish', 'private', 'future' ].indexOf( post.status ) === -1;
```
Might recommend splitting this long line:

```js
return (
	isEditedPostDirty( state ) ||
	hasMetaBoxes( state ) ||
	[ 'publish', 'private', 'future' ].indexOf( post.status ) === -1
);
```

Minor: I like semantic readability of `includes`:

```js
return (
	isEditedPostDirty( state ) ||
	hasMetaBoxes( state ) ||
	! includes( [ 'publish', 'private', 'future' ], post.status )
);
```

aduth commented on editor/store/test/selectors.js:
```diff
@@ -2086,6 +1996,21 @@ describe( 'selectors', () => {
 			expect( isSavingPost( state ) ).toBe( false );
 		} );
+
+		it( 'should return true if the post is notcurrently being saved but meta boxes are saving', () => {
```
Minor: Typo: Needs space between "not" and "currently"

aduth commented on editor/store/selectors.js:
```diff
@@ -65,32 +66,34 @@ export function getMetaBox( state, location ) {
-export const isMetaBoxStateDirty = ( state ) => getDirtyMetaBoxes( state ).length > 0;
+export const isSavingMetaBoxes = createSelector(
+	( state ) => {
+		return some( getMetaBoxes( state ), ( metaBox ) => {
```
Same note as previous.

aduth commented on editor/edit-post/meta-boxes.js:
```diff
@@ -0,0 +1,16 @@
+		return area.innerHTML;
+	}
+
+	return document.querySelector( '.metabox-location-' + location ).innerHTML;
```
When would we expect this line to be reached? (Can we add a comment explaining this?)

aduth replied:
And if we're expecting the first query selector might return an empty result, is there any reason to think this one might as well?

youknowriad replied:
The metaboxes are rendered to an initial hidden dom node (the later here) but when their panel is visible (it's not always the case for the sidebar), they are moved to their new location and when closed moved back to the original location.

aduth commented on editor/store/reducer.js:
```diff
@@ -668,6 +668,7 @@ export function metaBoxes( state = defaultMetaBoxState, action ) {
 					...state[ location ],
 					isLoaded: false,
 					isActive: action.metaBoxes[ location ],
+					html: getLocationHtml( location ),
```
Rather than calling into DOM from the reducer:

- Could the `MetaBoxesArea` component dispatch a "set HTML" action when it mounts?
- Should this at least be modeled as an effect? (Minding that with #3983 we should respect that effects occur after the original action)

aduth commented on editor/components/meta-boxes/meta-boxes-area/index.js:
```diff
@@ -84,30 +74,13 @@ class MetaBoxesArea extends Component {
-		const { loading } = this.state;
-		const { isDirty, changedMetaBoxState, location } = this.props;
-
-		const newIsDirty = ! isEqual( this.originalFormData, this.getFormData() );
```
Seems `this.originalFormData` is defunct now and can be removed, along with `getFormData`.

aduth reviewed (commented):
I don't love the DOM querying where we expect to know the class structure of meta box elements. The only idea I can think of as an alternative is a pattern similar to slots where dirtyable areas (components) can register themselves with `UnsavedChangesWarning`, and the initial HTML is tracked when the (dirtyable) component first mounts. At this point there'd be no need for the Redux property tracking.

Can you explain what it is about the changes here that resolves #4183?

youknowriad commented:
@aduth this closes #4183 because we never check for dirtyness before saving the metaboxes, each time we save a post, we save them. So by default the save button is visible.

aduth commented:
>@aduth this closes #4183 because we never check for dirtyness before saving the metaboxes, each time we save a post, we save them. So by default the save button is visible.

My reading of #4183 is not related to the Save button being unavailable, but rather that the default values aren't included in the payload of the save request.

youknowriad commented:
@aduth I think the request for saving the metaboxes is not triggered at all because they are considered "non-dirty"

youknowriad committed:
Meta Boxes: Changing the way we save metaboxes

Hidden metaboxes (side) were not saved before

youknowriad commented:
Ok, I think the PR is better now. I updated it. It still uses the DOM direct access. It's not great but since the metaboxes are kind of "weird chunk fo HTML" we have to deal with, it's acceptable IMO.
I added a comment to clarify its behavior.

Also,

 - I moved the save behavior of the metaboxes to an affect
 - If we have multiple areas, they're all saved together now fixing the bug where hidden metaboxes were not saved.
 - Some usage of `jQuery.serialize` to merge the different forms (I prefer this over having a custom code).

What do y'all think?

aduth commented on editor/components/meta-boxes/meta-boxes-area/index.js:
```diff
@@ -1,153 +1,85 @@
-				{ loading && <Spinner /> }
-				<div ref={ this.bindNode } />
+				{ isSaving && <Spinner /> }
+				<div className="editor-meta-boxes-area__container" ref={ this.bindNode } />
```
Unrelated: When I see `bindNode`, I assume it's assigning the `ref` of the root return value, not a child of the element. A minor thing, but I think it can be nice to clarify in the function name, e.g. `bindContainerNode`.

aduth commented on editor/components/meta-boxes/meta-boxes-area/index.js:
```diff
@@ -1,153 +1,85 @@
-		this.mounted = true;
-		this.fetchMetaboxes();
+		this.form = document.querySelector( '.metabox-location-' + this.props.location );
+		this.node.appendChild( this.form );
```
Not sure if we care to be guarded here, but if I pass an invalid `location` prop, this will error.

aduth commented on editor/store/actions.js:
```diff
@@ -438,11 +438,11 @@ export function removeNotice( id ) {
+ * this indicates that the sidebar has a meta box but the normal area does not.
  *
- * @param {Object} metaBoxes Whether meta box locations are active.
+ * @param   {Object} metaBoxes Whether meta box locations are active.
```
I think there was a faulty rebase here, I wouldn't expect these doc changes (i.e. we don't want the extra whitespace here, or in the `@returns` below)

aduth commented on editor/store/actions.js:
```diff
@@ -452,60 +452,38 @@ export function initializeMetaBoxState( metaBoxes ) {
+ * @returns {Object}                 Action object.
  */
-export function metaBoxStateChanged( location, hasChanged ) {
+export function metaBoxSetSavedData( dataPerLocation ) {
```
Unrelated: I prefer when the action creators are prefixed with the verb of what's happening, i.e. "set", `setMetaBoxSavedData`

aduth commented on editor/store/effects.js:
```diff
@@ -452,4 +455,38 @@ export default {
+			}
+			return memo;
+		}, {} );
+		store.dispatch( metaBoxSetSavedData( dataPerLocation ) );
```
Could be [DRY'd](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) up a bit with the `INITIALIZE_META_BOX_STATE` effect logic.

youknowriad replied:
Not so simple because the effect is triggering before the action. Didn't you make an update to `refx` to update this?

aduth replied:
>Not so simple because the effect is triggering before the action. Didn't you make an update to `refx` to update this?

Yes: #3983

aduth commented on editor/store/reducer.js:
```diff
@@ -661,60 +661,55 @@ const locations = [
+ * Reducer keeping track of the meta boxes saving state.
+ *
+ * @param {boolean} state  Previous state.
+ * @param {Object } action Action Object.
```
Extra space before the `}` is unexpected (here and in the DocBlock below)

aduth commented on editor/store/reducer.js:
```diff
@@ -661,60 +661,55 @@ const locations = [
+}
+
+/**
+ *
```
This line is unnecessary.

aduth commented on editor/store/reducer.js:
```diff
@@ -661,60 +661,55 @@ const locations = [
 }, {} );
 
+/**
+ * Reducer keeping track of the meta boxes saving state.
```
Personally I don't find this comment to be particularly helpful beyond pointing out that it's a reducer, and would encourage elaborating on the purpose for why we track a meta box saving state.

aduth commented on lib/meta-box-partial-page.php:
```diff
@@ -334,28 +304,29 @@ function the_gutenberg_metaboxes( $locations = array( 'advanced', 'normal', 'sid
-	}
+			</div>
+		</form>
+		<?php } ?>
```
I don't see a rule for it, but I'd conventionally have expected `if` and `endif` statements for conditions / loops within a PHP template:

>Note that requiring the use of braces just means that single-statement inline control structures are prohibited. You are free to use the alternative syntax for control structures (e.g. if/endif, while/endwhile)—especially in your templates where PHP code is embedded within HTML, for instance:
>
>```php
><?php if ( have_posts() ) : ?>
>    <div class="hfeed">
>        <?php while ( have_posts() ) : the_post(); ?>
>            <article id="post-<?php the_ID() ?>" class="<?php post_class() ?>">
>                <!-- ... -->
>            </article>
>        <?php endwhile; ?>
>    </div>
><?php endif; ?>
>```

https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#brace-style

aduth reviewed (commented):
I'm generally fine with the direction here.

youknowriad committed:
Rename bindNode and guard against empty nodes

youknowriad committed:
Rename setMetaBoxSavedData action

youknowriad committed:
Improve Meta Boxes reducer docs

youknowriad committed:
Use foreach and endforeach instead of curlybraces

youknowriad committed:
Fix dirty checking

youknowriad commented:
Any other concern here?

jeflopodev commented:
While using [Contextual Related Posts](https://github.com/WebberZone/contextual-related-posts) plugin I've noted I was able to **infinitely** save unmodified posts & pages.

After reporting them as a bug https://github.com/WebberZone/contextual-related-posts/issues/205 they linked back to this issue.

This feels bad for the UX / Workflow in my humble opinion. It gives me peace of mind when I see the button disabled. And the contrary creates me doubts, and insecurity of if I changed the last modification. I saw myself saving more than once, just in case......

This issue was from ~7 years ago. Many things may have changed.
So I'm wondering If Is there any possibility to reconsider how this works so the button can be disabled if there are no changes ?
Or if there's any other solution, that maybe the theme developer could apply to address this problem.

cc @youknowriad @aduth

youknowriad commented:
The best path forward for me is to move away from metaboxes. While we did ensure that metaboxes work in the block editor, they were and will never be first class citizen extensibility APIs. So I would recommend moving away from metaboxes in favor of regular block editor extensibility APIs. (for instance slots https://developer.wordpress.org/block-editor/reference-guides/slotfills/ ) and using the Data package to access and manipulate data.
